### ISSUE TITLE
Unhandled exception: Client is unable to create or apply a local media description

### ISSUE THREAD

**Issue by dambusm** [twilio/twilio-video.js#1504]
- [x] I have verified that the issue occurs with the latest twilio-video.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-video.js/blob/master/CHANGELOG.md).
 - [x] I reviewed the [Common Issues](https://github.com/twilio/twilio-video.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [x] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Software versions:**

 - [x] Browser(s): Edge 87.0.664
 - [x] Operating System: Windows 10
 - [x] twilio-video.js: 2.15.0
 - [x] Third-party libraries (e.g., Angular, React, etc.): React 17.0.2

**Issue description** 

I previously submitted an issue on this (#963) but the thread was closed as it was becoming too confused. I think it got conflated with this similarly named error: #1247.
Some users reported it was linked to other dependencies they were using (es6-promise-promise & webrtcHacks/adapter) but we are not using either of these.

The issue manifests as an unhandled exception, which we see quite often in our production logging:
```
TwilioError: Client is unable to create or apply a local media description
```

We are now capturing twilio-video's logging using the customized logger from 2.10.0, so can get a bit more context as to why it's thrown. It seems like the earliest sign of something going wrong is this:
```
"Failed to execute 'setRemoteDescription' on 'RTCPeerConnection': Failed to set remote offer sdp: The order of m-lines in subsequent offer doesn't match order from previous offer/answer.
```

Here is the full log from the most recent occurrence:

```
"[ '[TWILIO] Connecting to a Room',
    '[connect #1]',
    '2021-06-21T11:19:18.720Z' ]"
"[ '[TWILIO] LocalTracks were not provided, so they will be acquired automatically before connecting to the Room. LocalTracks will be released if connecting to the Room fails or if the Room is disconnected',
    '[connect #1]',
    '2021-06-21T11:19:18.721Z' ]"
"[ '[TWILIO] Call to getUserMedia successful; got MediaStreamTracks:',
    [ {} ],
    '[createLocalTracks #1]',
    '2021-06-21T11:19:18.833Z' ]"
"[ '[TWILIO] Created a new Participant',
    '[LocalParticipant #1]',
    '2021-06-21T11:19:18.837Z' ]"
"[ '[TWILIO] Added a new LocalAudioTrack:',
    '4c15bec6-7223-449f-9332-a220b32dc8c9',
    '[LocalParticipant #1]',
    '2021-06-21T11:19:18.838Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '131',
        group: 'signaling',
        level: 'info',
        name: 'early',
        timestamp: '1624274358850' },
    '[connect #1]',
    '2021-06-21T11:19:18.850Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '133',
        group: 'room',
        level: 'info',
        name: 'connect',
        timestamp: '1624274358852',
        payload: 
          { iceServers: '0',
            audioTracks: '0',
            videoTracks: '0',
            dataTracks: '0',
            audio: 'true',
            automaticSubscription: 'true',
            enableDscp: 'true',
            eventListener: 'false',
            preflight: 'false',
            video: 'false',
            enableDominantSpeaker: 'false',
            region: 'gll',
            preferredAudioCodecs: '[]',
            preferredVideoCodecs: '[]',
            networkQualityConfiguration: { remote: '0', local: '0' } } },
    '[connect #1]',
    '2021-06-21T11:19:18.852Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '1793',
        group: 'signaling',
        level: 'info',
        name: 'connecting',
        timestamp: '1624274360512' },
    '[connect #1]',
    '2021-06-21T11:19:20.512Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '2034',
        group: 'signaling',
        level: 'info',
        name: 'open',
        timestamp: '1624274360753' },
    '[connect #1]',
    '2021-06-21T11:19:20.753Z' ]"
"[ '[TWILIO] Added a new AudioTrackPublication:',
    'MTf111d42f5ad4c179950b0bc8cf4e5ea7',
    '[LocalParticipant #1]',
    '2021-06-21T11:19:21.514Z' ]"
"[ '[TWILIO] reconnected',
    '[LocalParticipant #1: PA21f1398e730ad8397a9e4fd381a6db46]',
    '2021-06-21T11:19:21.515Z' ]"
"[ '[TWILIO] Created a new Participant: caller',
    '[RemoteParticipant #2: PAa0268a24010e575b3d5baac02d5232a9]',
    '2021-06-21T11:19:21.517Z' ]"
"[ '[TWILIO] A new RemoteParticipant connected:',
    { audioTracks: {},
      dataTracks: {},
      identity: 'caller',
      networkQualityLevel: 'null',
      networkQualityStats: 'null',
      sid: 'PAa0268a24010e575b3d5baac02d5232a9',
      state: 'connected',
      tracks: {},
      videoTracks: {} },
    '[Room #1: RM6f6a7d85e0783103f451069d36a1aff7]',
    '2021-06-21T11:19:21.518Z' ]"
"[ '[TWILIO] Created a new Room:',
    'RM6f6a7d85e0783103f451069d36a1aff7',
    '[Room #1: RM6f6a7d85e0783103f451069d36a1aff7]',
    '2021-06-21T11:19:21.519Z' ]"
"[ '[TWILIO] Connected to Room:',
    '[Room #1: RM6f6a7d85e0783103f451069d36a1aff7]',
    '[connect #1]',
    '2021-06-21T11:19:21.519Z' ]"
"[ '[TWILIO] Room name:',
    'RM6f6a7d85e0783103f451069d36a1aff7',
    '[connect #1]',
    '2021-06-21T11:19:21.520Z' ]"
"[ '[TWILIO] Disabling',
    '[LocalAudioTrack #2: 4c15bec6-7223-449f-9332-a220b32dc8c9]',
    '2021-06-21T11:19:21.529Z' ]"
"[ '[TWILIO] Added a new DataTrack:',
    '7cba922a-72c6-41c6-9dc9-cfdd0af18ddf',
    '[LocalParticipant #1: PA21f1398e730ad8397a9e4fd381a6db46]',
    '2021-06-21T11:19:21.534Z' ]"
"[ '[TWILIO] Added a new LocalDataTrack:',
    '7cba922a-72c6-41c6-9dc9-cfdd0af18ddf',
    '[LocalParticipant #1: PA21f1398e730ad8397a9e4fd381a6db46]',
    '2021-06-21T11:19:21.535Z' ]"
"[ '[TWILIO] Enabling',
    '[LocalAudioTrack #2: 4c15bec6-7223-449f-9332-a220b32dc8c9]',
    '2021-06-21T11:19:21.546Z' ]"
"[ '[TWILIO] Added a new DataTrackPublication:',
    'MT4b7baca71ee0d735f8dff3ba56f507e4',
    '[LocalParticipant #1: PA21f1398e730ad8397a9e4fd381a6db46]',
    '2021-06-21T11:19:21.896Z' ]"
"[ '[TWILIO] Added a new AudioTrackPublication:',
    'MTf857a1be5f97bd7840d71d018a840da4',
    '[RemoteParticipant #2: PAa0268a24010e575b3d5baac02d5232a9]',
    '2021-06-21T11:19:22.063Z' ]"
"[ '[TWILIO] Added a new DataTrackPublication:',
    'MT7a419f1cb06e1762afc6324349d8481f',
    '[RemoteParticipant #2: PAa0268a24010e575b3d5baac02d5232a9]',
    '2021-06-21T11:19:22.064Z' ]"
"[ '[TWILIO] Added a new VideoTrackPublication:',
    'MT6d06231e267157916d3c13eb5d72cd63',
    '[RemoteParticipant #2: PAa0268a24010e575b3d5baac02d5232a9]',
    '2021-06-21T11:19:22.064Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '3491',
        group: 'media',
        level: 'warning',
        name: 'error',
        timestamp: '1624274362210',
        payload: 
          { message: 'Failed to _answer',
            code: '53400',
            context: '{\"error\":\"Failed to execute \\'setRemoteDescription\\' on \\'RTCPeerConnection\\': Failed to set remote offer sdp: The order of m-lines in subsequent offer doesn\\'t match order from previous offer/answer.\"}' } },
      '[connect #1]',
    '2021-06-21T11:19:22.210Z' ]"
"[ '[TWILIO] Failed to add RTCIceCandidate \"candidate:2580292971 1 udp 2122262783 2001:2061:2050:78a:14f2:623f:54f3:d2fd 50113 typ host generation 0 ufrag rAYy network-id 8 network-cost 900\": Failed to execute \\'addIceCandidate\\' on \\'RTCPeerConnection\\': Error processing ICE candidate',
      '[PeerConnectionV2 #1: d0e862db-57dc-43f2-aba7-aa79b81cab1d]',
    '2021-06-21T11:19:22.342Z' ]"
"[ '[TWILIO] Failed to add RTCIceCandidate \"candidate:2585960138 1 udp 2122000639 2001:2061:2050:78a:cd3:6860:de9d:6b1 49765 typ host generation 0 ufrag rAYy network-id 1 network-cost 50\": Failed to execute \\'addIceCandidate\\' on \\'RTCPeerConnection\\': Error processing ICE candidate',
      '[PeerConnectionV2 #1: d0e862db-57dc-43f2-aba7-aa79b81cab1d]',
    '2021-06-21T11:19:22.342Z' ]"
"[ '[TWILIO] Failed to add RTCIceCandidate \"candidate:2585960138 1 udp 2121935103 2001:2061:2050:78a:cd3:6860:de9d:6b1 59934 typ host generation 0 ufrag rAYy network-id 4 network-cost 50\": Failed to execute \\'addIceCandidate\\' on \\'RTCPeerConnection\\': Error processing ICE candidate',
      '[PeerConnectionV2 #1: d0e862db-57dc-43f2-aba7-aa79b81cab1d]',
    '2021-06-21T11:19:22.343Z' ]"
"[ '[TWILIO] Failed to add RTCIceCandidate \"candidate:2200656040 1 udp 1685987071 46.132.21.102 16389 typ srflx raddr 0.0.0.0 rport 0 generation 0 ufrag rAYy network-id 7 network-cost 900\": Failed to execute \\'addIceCandidate\\' on \\'RTCPeerConnection\\': Error processing ICE candidate',
      '[PeerConnectionV2 #1: d0e862db-57dc-43f2-aba7-aa79b81cab1d]',
    '2021-06-21T11:19:22.343Z' ]"
"[ '[TWILIO] Failed to add RTCIceCandidate \"candidate:2686458394 1 udp 41820159 18.195.48.247 52302 typ relay raddr 46.132.21.102 rport 16389 generation 0 ufrag rAYy network-id 7 network-cost 900\": Failed to execute \\'addIceCandidate\\' on \\'RTCPeerConnection\\': Error processing ICE candidate',
      '[PeerConnectionV2 #1: d0e862db-57dc-43f2-aba7-aa79b81cab1d]',
    '2021-06-21T11:19:22.343Z' ]"
"[ '[TWILIO] Failed to add RTCIceCandidate \"candidate:4029852575 1 udp 8265471 18.195.48.247 45367 typ relay raddr 46.132.21.102 rport 16518 generation 0 ufrag rAYy network-id 7 network-cost 900\": Failed to execute \\'addIceCandidate\\' on \\'RTCPeerConnection\\': Error processing ICE candidate',
      '[PeerConnectionV2 #1: d0e862db-57dc-43f2-aba7-aa79b81cab1d]',
    '2021-06-21T11:19:22.665Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '12796',
        group: 'quality',
        level: 'info',
        name: 'stats-report',
        timestamp: '1624274371515',
        payload: 
          { audioTrackStats: [],
            localAudioTrackStats: 
            [ { trackId: '5c6e8860-5cab-4288-8497-3fba7a3d64ab',
                trackSid: 'MTf111d42f5ad4c179950b0bc8cf4e5ea7',
                timestamp: '1624274371514',
                ssrc: '551804638',
                packetsLost: 'null',
                codec: 'opus',
                bytesSent: '0',
                packetsSent: '0',
                roundTripTime: '0',
                audioLevel: 'null',
                jitter: 'null' } ],
          localVideoTrackStats: [],
          peerConnectionId: 'd0e862db-57dc-43f2-aba7-aa79b81cab1d',
          videoTrackStats: [] } },
    '[connect #1]',
    '2021-06-21T11:19:31.515Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '12798',
        group: 'quality',
        level: 'info',
        name: 'active-ice-candidate-pair',
        timestamp: '1624274371517',
        payload: 
          { availableIncomingBitrate: '0',
            availableOutgoingBitrate: '0',
            bytesReceived: '0',
            bytesSent: '0',
            consentRequestsSent: '0',
            currentRoundTripTime: '0',
            lastPacketReceivedTimestamp: '0',
            lastPacketSentTimestamp: '0',
            nominated: 'false',
            peerConnectionId: 'd0e862db-57dc-43f2-aba7-aa79b81cab1d',
            priority: '0',
            readable: 'false',
            requestsReceived: '0',
            requestsSent: '0',
            responsesReceived: '0',
            responsesSent: '0',
            retransmissionsReceived: '0',
            retransmissionsSent: '0',
            state: 'failed',
            totalRoundTripTime: '0',
            transportId: '',
            writable: 'false',
            localCandidate: 
            { candidateType: 'host',
              deleted: 'false',
              ip: '',
              port: '0',
              priority: '0',
              protocol: 'udp',
              url: '' },
          remoteCandidate: 
          { candidateType: 'host',
              ip: '',
              port: '0',
              priority: '0',
              protocol: 'udp',
              url: '' } } },
    '[connect #1]',
    '2021-06-21T11:19:31.517Z' ]"
"[ '[TWILIO] event',
    { elapsedTime: '18454',
        group: 'media',
        level: 'warning',
        name: 'error',
        timestamp: '1624274377173',\n    payload: \n     { message: 'Failed to _answer',\n       code: '53400',\n       context: '{\"error\":\"Failed to execute \\'setRemoteDescription\\' on \\'RTCPeerConnection\\': Failed to set remote offer sdp: The order of m-lines in subsequent offer doesn\\'t match order from previous offer/answer.\"}' } },\n  '[connect #1]',\n  '2021-06-21T11:19:37.173Z' ]"
"TwilioError: Client is unable to create or apply a local media description"
```

**Comment by makarandp0** [comment]
Hello @dambusm,  Thank you for reporting this issue and detailed logs. I took a quick look at the logs and looks like this was a peer to peer room with two participants one using mobile sdk that uses planB and other using javascript sdk (unified plan). 

```
P1: PAa0268a24010e575b3d5baac02d5232a9 | CAcd8179f24d62e2264933db2ce63af6c6 | caller | planb | ios 4.3.0 / 15.0 
P2: PA21f1398e730ad8397a9e4fd381a6db46 | CAd8ae395e61161195e166c6ffeb7b26e4 | responder | unified | jsdk 2.15.0 | Edge 87.0.664 
```

I will look further into this, but I have a quick question -  I am wondering if this is typical for other cases you are seeing from your customers. Do all these failures involve interop between planB and unified plan?

Thanks,
Makarand

**Comment by dambusm** [comment]
Hi @makarandp0 , yes, all of our calls are initiated on native mobile apps using the iOS & android sdks and answered on web using the js sdk. From the looks of it, unified plan is only available in the latest iOS sdk beta release, which we're not using in production, so I guess all calls are initiated with planB.

**Comment by makarandp0** [comment]
Hey @dambusm, sorry for delay on getting back to you - we are aware of some plan-B and unified plan interop issues and are tracking them with internal ticket (VIDEO-5958). I will update as we find more on those. In the meanwhile if you find repro steps to consistently reproduce the issue, please let us know.

**Comment by shaibt** [comment]
We're seeing the same issues on iOS side of P2P calls when using `TwilioVideo` ver 4.6 (latest iOS version) and Chrome browser (on OSX) using `twilio-video-js` ver 2.13.1.
Chrome browser side video and audio is not received on iOS (intermittently) and logs show `Client is unable to create or apply a local media description`

**Comment by makarandp0** [comment]
Hello @shaibt, Can you please share room-sid and client side logs?  
Thanks,
Makarand

**Comment by shaibt** [comment]
Hi @makarandp0 here is a room-sid: `RM937e2538aba8a05adfd565fa89c27a16`
Also log file for the iOS side that includes the errors - cannot view JS published channels on the iOS end.

[twilio.log](https://github.com/twilio/twilio-video.js/files/7288540/twilio.log)

**Comment by ceaglest** [comment]
Hi @shaibt,

> @makarandp0 here is a room-sid: RM937e2538aba8a05adfd565fa89c27a16

Thank you for providing this example. From looking at the SDK and Server logs the issue may be related to the use of dynamic payload types in the [35,63] range and the iOS SDK's handling of that. 

> Chrome browser side video and audio is not received on iOS (intermittently) and logs show Client is unable to create or apply a local media description.

I tried to reproduce the same scenario locally but was not successful yet, so I am curious how reproducible it is for you? It seems to be important to have the second participant be Chrome and for that participant to have enough video codecs available to exceed the 127 dynamic PT range.

Best,
Chris

**Comment by ceaglest** [comment]
Hi,

For anyone reading recent comments in this issue, we landed a fix on the iOS side in 4.6.1. See https://github.com/twilio/twilio-video-ios/issues/211 for details.

Best,
Chris

**Comment by makarandp0** [comment]
Thank you @ceaglest. Closing this issue now.
